ambry-network/src/main/java/com/github/ambry/network/NettyServerRequestResponseChannel.java

```diff
@@ -80,30 +104,91 @@ public void closeConnection(NetworkRequest originalRequest) throws InterruptedEx
+      }
+      requestToServe = networkRequestBundle.getRequestToServe();
+      requestsToDrop = networkRequestBundle.getRequestsToDrop();
+
+      // Consume the first 8 bytes of requests since TCP impl uses this size to allocate buffer.
+      if (requestToServe != null) {
+        http2ServerMetrics.requestQueuingTime.update(System.currentTimeMillis() - requestToServe.getStartTimeInMs());
+        if (!consumeStream(requestToServe)) {
+          requestToServe = null;
+        }
+      }
+      Set<NetworkRequest> errorRequests = new HashSet<>();
+      for (NetworkRequest requestToDrop : requestsToDrop) {
+        http2ServerMetrics.requestQueuingTime.update(System.currentTimeMillis() - requestToDrop.getStartTimeInMs());
+        if (!consumeStream(requestToDrop)) {
+          errorRequests.add(requestToDrop);
+          logger.error("Error reading the stream, dropping the request: {}", requestToDrop);
+        }
+      }
+      requestsToDrop.removeAll(errorRequests);
+    } while (requestToServe == null && requestsToDrop.isEmpty());
+
+    return new NetworkRequestBundle(requestToServe, requestsToDrop);
+  }
+
   @Override
-  public NetworkRequest receiveRequest() throws InterruptedException {
-
-    NetworkRequest request = null;
-    while (request == null) {
-      request = requestQueue.take();
-      http2ServerMetrics.requestQueuingTime.update(System.currentTimeMillis() - request.getStartTimeInMs());
-      if (request.equals(EmptyRequest.getInstance())) {
-        logger.debug("Request handler {} received shut down command ", request);
-      } else {
-        DataInputStream stream = new DataInputStream(request.getInputStream());
-        try {
-          // The first 8 bytes is size of the request. TCP implementation uses this size to allocate buffer. See {@link BoundedReceive}
-          // Here we just need to consume it.
-          stream.readLong();
-        } catch (IOException e) {
-          logger.error("Encountered an error while reading length out, close the connection", e);
-          closeConnection(request);
-          request = null;
+  public List<NetworkRequest> getUnqueuedRequests() throws InterruptedException {
+    List<NetworkRequest> requestsToDrop = new ArrayList<>();
+    NetworkRequest unqueuedRequest;
+    do {
+      while ((unqueuedRequest = unqueuedRequests.poll()) != null) {
+        requestsToDrop.add(unqueuedRequest);
+      }
+      if (requestsToDrop.isEmpty()) {
+        // If there are no un-queued requests, wait on the queue.
+        unqueuedRequest = unqueuedRequests.take();
+        requestsToDrop.add(unqueuedRequest);
+      }
+      Set<NetworkRequest> errorRequests = new HashSet<>();
+      for (NetworkRequest requestToDrop : requestsToDrop) {
+        if (!consumeStream(requestToDrop)) {
+          errorRequests.add(requestToDrop);
+          logger.error("Error reading the stream, dropping the request: {}", requestToDrop);
```

Comment:
i feel like this loop can be simplified. 

`  @Override
  public List<NetworkRequest> getUnqueuedRequestsX() throws InterruptedException {
    List<NetworkRequest> requestsToDrop = new ArrayList<>();
    NetworkRequest unqueuedRequest;
    while (requestsToDrop.isEmpty()) {
      while ((unqueuedRequest = unqueuedRequests.poll()) != null) {
        if (consumeStream(requestToDrop)) {
          requestsToDrop.add(unqueuedRequest);
        }
      }
      return requestsToDrop;
    }
  }`

Reply:
if requestsToDrop is empty, will a call to this method block ?

Reply:
Thank you for suggestion. Simplified the loop.

> or why not let just dropRequest handle what to do with these errorneous requests instead of branching here ?

I think we need to keep code to read first 8 bytes at this layer so that it can be applicable to both valid and dropped requests. 

> if requestsToDrop is empty, will a call to this method block ?

Yeah, this method will block if requestsToDrop is empty. Added it as comments in javadoc.